README.md
# react-engine

[![Build Status](https://travis-ci.org/paypal/react-engine.svg?branch=master)](https://travis-ci.org/paypal/react-engine)

### What is react-engine?
* a react render engine for [Universal](https://medium.com/@mjackson/universal-javascript-4761051b7ae9) (previously [Isomorphic](http://nerds.airbnb.com/isomorphic-javascript-future-web-apps/)) JavaScript apps written with express
* renders both plain react views and **optionally** react-router views
* enables server rendered views to be client mountable


### Install
```sh
# In your express app, react-engine needs to be installed alongside react/react-dom (react-router is optional)
$ npm install react-engine react react-dom react-router --save
```

### Usage On Server Side
###### Setup in an Express app
```javascript
var Express = require('express');
var ReactEngine = require('react-engine');

var app = Express();

// create an engine instance
var engine = ReactEngine.server.create({
  /*
    see the complete server options spec here:
    https://github.com/paypal/react-engine#server-options-spec
  */
});

// set the engine
app.engine('.jsx', engine);

// set the view directory
app.set('views', __dirname + '/views');

// set jsx or js as the view engine
// (without this you would need to supply the extension to res.render())
// ex: res.render('index.jsx') instead of just res.render('index').
app.set('view engine', 'jsx');

// finally, set the custom view
app.set('view', require('react-engine/lib/expressView'));
```

###### Setup in a [KrakenJS](http://krakenjs.com) app's config.json
```json
{
  "express": {
    "view engine": "jsx",
    "view": "require:react-engine/lib/expressView",
  },
  "view engines": {
    "jsx": {
      "module": "react-engine/lib/server",
      "renderer": {
        "method": "create",
        "arguments": [{
          /*
            see the complete server options spec here:
            https://github.com/paypal/react-engine#server-options-spec
          */
        }]
      }
    }
  }
}
```

###### Server options spec
Pass in a JavaScript object as options to the react-engine's [server engine create method](#setup-in-an-express-app).
The options object should contain the mandatory `routes` property with the route definition.

Additionally, it can contain the following **optional** properties,

- `docType`: \<String> - a string that can be used as a doctype (_Default: `<!DOCTYPE html>`_).
                        (docType might not make sense if you are rendering partials/sub page components, in that case you can pass an empty string as docType)
- `routesFilePath`: \<String> - path for the file that contains the react router routes.
                   react-engine uses this behind the scenes to reload the routes file in
                   cases where [express's app property](http://expressjs.com/api.html#app.set) `view cache` is false, this way you don't need to restart the server every time a change is made in the view files or routes file.
- `renderOptionsKeysToFilter`: \<Array> - an array of keys that need to be filtered out from the data object that gets fed into the react component for rendering. [more info](#data-for-component-rendering)
- `performanceCollector`: \<Function> - to collects [perf stats](#performance-profiling)
- `scriptLocation`: \<String> - where in the HTML you want the client data (i.e. `<script>var __REACT_ENGINE__ = ... </script>`) to be appended (_Default: `body`_).
                    If the value is undefined or set to `body` the script is placed before the `</body>` tag.
                    The only other value is `head` which appends the script before the `</head>` tag.

###### Rendering views on server side
```js
var data = {}; // your data model

// for a simple react view rendering
res.render(viewName, data);

// for react-router rendering
// pass in the `url` and react-engine
// will run the react-router behind the scenes.
res.render(req.url, data);
```

### Usage On Client Side (Mounting)
```js
// assuming we use `browserify`
var client = require('react-engine').client;

// finally, boot whenever your app is ready
// example:
document.addEventListener('DOMContentLoaded', function onLoad() {

  // `onBoot` - Function (optional)
  // returns data that was used
  // during rendering as the first argument
  // the second argument is the `history` object that was created behind the scenes
  // (only available while using react-router)
  client.boot(/* client options object */, function onBoot(data, history) {

  });
};

// if the data is needed before booting on
// client, call `data` function anytime to get it.
// example:
var data = client.data();
```

###### Client options spec
Pass in a JavaScript object as options to the react-engine's client boot function.
It can contain the following properties,

- `routes` : **required** - _Object_ - the route definition file.
- `viewResolver` : **required** - _Function_ - a function that react-engine needs to resolve the view file.
  an example of the viewResolver can be [found here](https://github.com/paypal/react-engine/blob/ecd27b30a9028d3f02b8f8e89d355bb5fc909de9/examples/simple/public/index.js#L29).
- `mountNode` : **optional** - _HTMLDOMNode_ - supply a HTML DOM Node to mount the server rendered component in the case of partial/non-full page rendering.
- `history` : **optional** - _Object_ - supply any custom history object to be used by the react-router.

### Data for component rendering
The actual data that gets fed into the component for rendering is the `renderOptions` object that [express generates](https://github.com/strongloop/express/blob/2f8ac6726fa20ab5b4a05c112c886752868ac8ce/lib/application.js#L535-L588).

If you don't want to pass all that data, you can pass in an array of object keys that react-engine can filter out from the `renderOptions` object before passing it into the component for rendering.

```javascript
// example of using `renderOptionsKeysToFilter` to filter `renderOptions` keys
var engine = ReactEngine.server.create({
  renderOptionsKeysToFilter: ['mySensitiveDataThatIsIn_res.locals'],
  routes: require(path.join(__dirname + './reactRoutes'))
});
```

Note: By default, the following three keys are always filtered out from `renderOptions` no matter whether `renderOptionsKeysToFilter` is configured or not.

- `settings`
- `enrouten`
- `_locals`

### Handling redirects and route not found errors on the server side
While using react-router, it matches the url to a component based on the app's defined routes. react-engine captures the redirects and not-found cases that are encountered while trying to run the react-router's [match function on the server side](https://github.com/reactjs/react-router/blob/master/docs/guides/ServerRendering.md).

To handle the above during the lifecycle of a request, add an error type check in your express error middleware. The following are the three types of error that get thrown by react-engine:

Error Type           | Description   
-------------------- | --------------------------------------------------------
MATCH_REDIRECT**     | indicates that the url  matched to a redirection
MATCH_NOT_FOUND      |  indicates that the url  did not match to any component     
MATCH_INTERNAL_ERROR | indicates that react-router encountered an internal error

 _**  for MATCH_REDIRECT error, `redirectLocation` property of the err has the new redirection location_

```javascript
// example express error middleware
app.use(function(err, req, res, next) {
  console.error(err);

  // http://expressjs.com/en/guide/error-handling.html
  if (res.headersSent) {
    return next(err);
  }

  if (err._type && err._type === ReactEngine.reactRouterServerErrors.MATCH_REDIRECT) {
    return res.redirect(302, err.redirectLocation);
  }
  else if (err._type && err._type === ReactEngine.reactRouterServerErrors.MATCH_NOT_FOUND) {
    return res.status(404).send('Route Not Found!');
  }
  else {
    // for ReactEngine.reactRouterServerErrors.MATCH_INTERNAL_ERROR or
    // any other error we just send the error message back
    return res.status(500).send(err.message);
  }
});
```

### Yeoman Generator
There is a Yeoman generator available to create a new express or KrakenJS application which uses react-engine:
[generator-react-engine](https://www.npmjs.com/package/generator-react-engine).

### Performance Profiling

Pass in a function to the `performanceCollector` property to collect the `stats`
object for every render.

##### `stats`
The object that contains the stats info for each render by react-engine.
It has the below properties.
- `name` - Name of the template or the url in case of react router rendering.
- `startTime` - The start time of render.
- `endTime` - The completion time of render.
- `duration` - The duration taken to render (in milliseconds).

```js
// example
function collector(stats) {
  console.log(stats);
}

var engine = require('react-engine').server.create({
  routes: './routes.jsx'
  performanceCollector: collector
});
```

### Notes
* On the client side, the state is exposed on the window object's property `__REACT_ENGINE__`
* When Express's `view cache` app property is false (mostly in non-production environments), views are automatically reloaded before render. So there is no need to restart the server for seeing the changes.
* You can use `js` as the engine if you decide not to write your react views in `jsx`.
* You can add [nonce](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#Unsafe_inline_script) in `_locals`, which will be added in `script` tag that gets injected into the server rendered pages, like `res.locals.nonce = 'nonce value'` 
* [Blog on react-engine](https://www.paypal-engineering.com/2015/04/27/isomorphic-react-apps-with-react-engine/)


### License
[Apache Software License v2.0](http://www.apache.org/licenses/LICENSE-2.0)


CHANGELOG.md
## 4.4.0
* Add support to React 16
* Update peerDependencies to remove installation warnings
* Update devDependencies to execute test in React 16

## 4.3.0 (Feb 7 2017)
* Add option to change injected script type from the default 'application/json'.

## 4.2.1 (Dec 9 2016)
* Converted from escape-html to jsesc  (https://github.com/paypal/react-engine/pull/185)

## 4.2.0 (Oct 24 2016)
* XSS safe implementation of prop passing from server to client (https://github.com/paypal/react-engine/pull/179/)
* Specify `react` and `react-dom` as peer dependencies (https://github.com/paypal/react-engine/pull/177)
* replace full lodash inlined into client with lodash/assign (https://github.com/paypal/react-engine/pull/176)
* support dot-lookup paths in renderOptionsKeysToFilter (https://github.com/paypal/react-engine/pull/175)

## 4.1.0 (Aug 11 2016)
* safeguard against property over look while fusing together data object and routerProps object in the server render (https://github.com/paypal/react-engine/pull/173/)

## 4.0.0 (July 1 2016)
* removed `react-dom` from being a `react-engine` dependency
* updated a lot of dependencies and fixed tests
* remove JSCS/grunt and added airbnb eslint config

## 3.4.1 (May 16 2016)
* add backward compatibility for react-router@1 (https://github.com/paypal/react-engine/pull/159)

## 3.4.0 (May 13 2016)
* Update deprecated `history` and `RoutingContext` for react-router (https://github.com/paypal/react-engine/pull/155)

## 3.3.0 (Apr 30 2016)
* Added scriptLocation server option to allow consumers to specify location of REACT_ENGINE script (https://github.com/paypal/react-engine/pull/153)
* Support ES6 module syntax for routes (https://github.com/paypal/react-engine/pull/154)

## 3.2.2 (Apr 19 2016)

* fix #151, make react-router optional (https://github.com/paypal/react-engine/pull/152)

## 3.2.1 (Apr 12 2016)

* Support ES6 module syntax for React views (https://github.com/paypal/react-engine/pull/149)

## 3.2.0 (Mar 27 2016)

* Allow consumers to override history object  (https://github.com/paypal/react-engine/issues/126)

## 3.1.0 (Jan 25 2016)

* fix - set implicit extension to import file names
* fix - Allow consumers to override history object
* Use path instead of pathname to ensure querystring is not stripped - https://github.com/paypal/react-engine/pull/131
* Client-side error when using code splitting in webpack - https://github.com/paypal/react-engine/pull/129

## 3.0.0 (Jan 10 2016)

* [v3.x] - support react-router@1 and react@0.14

## 2.6.2 (Jan 3 2016)

* fix - lodash-node package is deprecated(https://github.com/paypal/react-engine/issues/122)

## 2.6.1 (Dec 30 2015)

* fix undefined createOptions var [client.js] (https://github.com/paypal/react-engine/issues/119)

## 2.6.0 (Nov 06 2015)

* make the render root configurable (https://github.com/paypal/react-engine/issues/68)

## 2.5.0 (Oct 29 2015)

* Throw an error only if peer dependency is not installed and is really required (https://github.com/paypal/react-engine/pull/98)

## 2.4.0 (Oct 15 2015)

* Export Router object to consumers. (https://github.com/paypal/react-engine/issues/81)

## 2.3.0 (Oct 11 2015)

* Allow custom doctype option. (https://github.com/paypal/react-engine/pull/96)

## 2.2.1 (Oct 09 2015)

* make the clearRequireCacheInDir platform windows friendly (https://github.com/paypal/react-engine/issues/93)

## 2.2.0 (Sep 02 2015)

* Allow finer grain control of render properties (https://github.com/paypal/react-engine/issues/73)

## 2.1.0 (Aug 20 2015)

* resolve cache clear logic based on the 'view cache' (https://github.com/paypal/react-engine/issues/74)
* updated readme with migration to v2.x notes (https://github.com/paypal/react-engine/issues/75)
* updated readme references of Isomorphic JavaScript to Universal JavaScript (https://github.com/paypal/react-engine/issues/60)

## 2.0.0 (Aug 1 2015)

* Major API changes (specifically the options object property name changes)
* React-Router config properties can be passed through the react engine now.

## 1.7.0 (June 22, 2015)

* Windows path fix (https://github.com/paypal/react-engine/pull/41)

## 1.6.0 (May 13, 2015)

* expose state/data on the client side using additional function called data. Helps in flux implementations, which need data even before booting.

## 1.5.0 (May 9, 2015)

* made peerDependencies and dependencies, `react` and `react-router`'s versions to be more flexible.

## 1.4.1 (May 7, 2015)

* Fix: https://github.com/paypal/react-engine/issues/28
* add unit tests for expressView.js
* change tape test reporter from `tap-spec` to `faucet`

## 1.4.0 (May 3, 2015)

* remove passing react & react-router as options in the client boot.
* lock down version of `jsdom` (latest versions seem to fail tests in Node.js env)

## 1.3.0 (April 30, 2015)

* added performance profiling

## 1.2.0 (April 25, 2015)

* generate semantic html by injecting script tag before end of html tag
* https://github.com/paypal/react-engine/pull/16

## 1.1.0 (April 11, 2015)

* added an API to the client side code to expose data.
* added ChangeLog and .npmignore
* added tap-spec to pretty format tape test results
* added .editorconfig file

## 1.0.0 (April 9, 2015)

* initial release


lib/server.js
/*-------------------------------------------------------------------------------------------------------------------*\
|  Copyright (C) 2017 PayPal                                                                                           |
|                                                                                                                     |
|  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance     |
|  with the License.                                                                                                  |
|                                                                                                                     |
|  You may obtain a copy of the License at                                                                            |
|                                                                                                                     |
|       http://www.apache.org/licenses/LICENSE-2.0                                                                    |
|                                                                                                                     |
|  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed   |
|  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for  |
|  the specific language governing permissions and limitations under the License.                                     |
\*-------------------------------------------------------------------------------------------------------------------*/

'use strict';

var isString = require('lodash/isString');
var assign = require('lodash/assign');
var unset = require('lodash/unset');
var path = require('path');
var util = require('./util');
var assert = require('assert');
var Config = require('./config.json');
var jsesc = require('jsesc');
var ReactDOMServer = require('react-dom/server');
var debug = require('debug')(require('../package').name);
var ReactRouterServerErrors = require('./reactRouterServerErrors');

var format = require('util').format;
var Performance = require('./performance');

// safely require the peer-dependencies
var React = util.safeRequire('react');

function generateReactRouterServerError(type, existingErrorObj, additionalProperties) {
  var err = existingErrorObj || new Error('react router match fn error');
  err._type = type;
  if (additionalProperties) {
    assign(err, additionalProperties);
  }

  return err;
}

exports.create = function create(createOptions) {
  createOptions = createOptions || {};

  // safely require the peer-dependencies
  var React = util.safeRequire('react');
  var Router;
  var match;
  var RouterContext;

  try {
    Router = require('react-router');
    match = Router.match;

    // compatibility for both `react-router` v2 and v1
    RouterContext = Router.RouterContext || Router.RoutingContext;
  } catch (err) {
    if (!Router && createOptions.routes) {
      throw err;
    }
  }

  createOptions.scriptType = isString(createOptions.scriptType) ? createOptions.scriptType : Config.scriptType;
  createOptions.docType = isString(createOptions.docType) ? createOptions.docType : Config.docType;
  createOptions.renderOptionsKeysToFilter = createOptions.renderOptionsKeysToFilter || [];
  createOptions.staticMarkup = createOptions.staticMarkup !== undefined ? createOptions.staticMarkup : Config.staticMarkup;
  createOptions.styledComponents = createOptions.styledComponents !== undefined ? createOptions.styledComponents : Config.styledComponents;

  assert(Array.isArray(createOptions.renderOptionsKeysToFilter),
      '`renderOptionsKeysToFilter` - should be an array');

  createOptions.renderOptionsKeysToFilter =
    createOptions.renderOptionsKeysToFilter.concat(Config.defaultKeysToFilter);

  if (createOptions.performanceCollector) {
    assert.equal(typeof createOptions.performanceCollector,
        'function',
        '`performanceCollector` - should be a function');
  }

  // the render implementation
  return function render(thing, options, callback) {

    var perfInstance;

    if (createOptions.performanceCollector) {
      perfInstance = Performance(thing);
    }

    function done(err, html) {
      if (!options.settings['view cache']) {
        // remove all the files under the express's view folder from require cache.
        // Helps in making changes to react views without restarting the server.
        util.clearRequireCache(createOptions.routesFilePath);
        util.clearRequireCacheInDir(options.settings.views, options.settings['view engine']);
      }

      if (createOptions.performanceCollector) {
        createOptions.performanceCollector(perfInstance());
      }

      callback(err, html);
    }

    function renderAndDecorate(component, data, html) {
      if (createOptions.staticMarkup) {
        // render the component to static markup
        return ReactDOMServer.renderToStaticMarkup(component);
      }

      // render the redux wrapped component
      if (createOptions.reduxStoreInitiator) {
        // add redux provider
        var Provider = require('react-redux').Provider;
        var initStore;
        try {
          initStore = require(createOptions.reduxStoreInitiator);
          if (initStore.default) {
            initStore = initStore.default;
          }
          var store = initStore(data);
          component = React.createElement(Provider, { store: store }, component);
        } catch (err) {
          return done(err);
        }
      }

      // render the component and get styled-components styles
      if (createOptions.styledComponents) {
        var ServerStyleSheet = require('styled-components').ServerStyleSheet;
        var sheet = new ServerStyleSheet();
        try {
          html += ReactDOMServer.renderToString(sheet.collectStyles(component));
          var styleTags = sheet.getStyleTags();
          // add the styles to the end of the head
          var htmlTag = '</head>';
          html = html.replace(htmlTag, styleTags + htmlTag);
        } catch (err) {
          return done(err);
        }
      } else {
        html += ReactDOMServer.renderToString(component);
      }

      // the `script` tag that gets injected into the server rendered pages.
      // https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet#RULE_.233_-_JavaScript_Escape_Before_Inserting_Untrusted_Data_into_JavaScript_Data_Values
      var openScriptTag = `<script id="${Config.client.markupId}" type="${createOptions.scriptType}" ${options.nonce ? `nonce="${options.nonce}"` : ''}>`;
      // Escape data for injecting into <script> tag
      // https://mathiasbynens.be/notes/etago
      var script = openScriptTag + jsesc(data, {
            'escapeEtago': true, // old option for escaping in <script> or <style> context
            'isScriptContext': true, // soon to be new option
            'compact': true, // minifies
            'json': true // ensures JSON compatibility
          })
        + '</script>';


      if (createOptions.docType === '') {
        // if the `docType` is empty, the user did not want to add a docType to the rendered component,
        // which means they might not be rendering a full page with `html` and `body` tags
        // so attach the script tag to just  the end of the generated html string
        html += script;
        html += styleTags;
      } else {
        var htmlTag = createOptions.scriptLocation === 'head' ? '</head>' : '</body>';
        html = html.replace(htmlTag, script + htmlTag);
      }

      return html;
    }

    if (createOptions.routes && createOptions.routesFilePath) {
      // if `routesFilePath` property is provided, then in
      // cases where 'view cache' is false, the routes are reloaded for every render.
      createOptions.routes = require(createOptions.routesFilePath);
      if (createOptions.routes.default) {
        createOptions.routes = createOptions.routes.default;
      }
    }

    // initialize the markup string
    var html = createOptions.docType;

    // create the data object that will be fed into the React render method.
    // Data is a mash of the express' `render options` and `res.locals`
    // and meta info about `react-engine`
    var data = assign({
      __meta: {
        // get just the relative path for view file name
        view: null,
        markupId: Config.client.markupId
      }
    }, options);
    if (this.useRouter && !createOptions.routes) {
      return done(new Error('asking to use react router for rendering, but no routes are provided'));
    }

    // since `unset` mutates the obj, lets clone a copy
    // Also, we are using JSON.parse(JSON.stringify(data)) to clone the object super fast.
    // a valid assumption in using this method of cloning at this point: we have only variables
    // and not any functions in data object - so need for lodash cloneDeep
    try {
      data = JSON.parse(JSON.stringify(data));
      createOptions.renderOptionsKeysToFilter.forEach(function(key) {
        unset(data, key);
      });
    } catch (parseErr) {
      return done(parseErr);
    }

    try {
      if (this.useRouter) {
        return match({ routes:createOptions.routes, location:thing}, function reactRouterMatchHandler(error, redirectLocation, renderProps) {
          if (error) {
            debug('server.js match 500 %s', error.message);
            var err = generateReactRouterServerError(ReactRouterServerErrors.MATCH_INTERNAL_ERROR, error);
            return done(err);
          } else if (redirectLocation) {
            debug('server.js match 302 %s', redirectLocation.pathname + redirectLocation.search);
            var err = generateReactRouterServerError(ReactRouterServerErrors.MATCH_REDIRECT, null, {
              redirectLocation: redirectLocation.pathname + redirectLocation.search
            });
            return done(err);
          } else if (renderProps) {
            renderProps.createElement = function(Component, routerProps) {
              // Other than fusing the data object with the routerProps, there is no way
              // to pass data into the routing context of react-router during a server render.
              // since we are going to use `assign` to fuse the routerProps and the actual
              // data object, we need to make sure that there are no properties between the two object
              // with the same name at the root level. (Having two properties with the same name breaks assign.)
              // Info on why we need to fuse the two objects?
              // --------------------------------------------
              // * https://github.com/ngduc/react-setup/issues/10
              // * https://github.com/reactjs/react-router/issues/1969
              // * http://stackoverflow.com/questions/36137901/react-route-and-server-side-rendering-how-to-render-components-with-data
              if (options.settings.env !== 'production') {
                var intersection = Object.keys(routerProps).filter(function(elem) {
                  return Object.keys(data).indexOf(elem) !== -1;
                });
                if (intersection.length) {
                  var errMsg = 'Your data object cannot have property(ies) named: "' +
                    intersection +
                    '"\n Blacklisted property names that cannot be used: "' +
                    Object.keys(routerProps) +
                    '"\n'
                  throw new Error(errMsg);
                }
              }

              // define a createElement strategy for react-router that transfers data props to all route "components"
              // for any component created by react-router, fuse data object with the routerProps
              // NOTE: This may be imposing too large of an opinion?
              return React.createElement(Component, assign({}, data, routerProps));
            };

            return done(null, renderAndDecorate(React.createElement(RouterContext, renderProps), data, html));
          } else {
            debug('server.js match 404');
            var err = generateReactRouterServerError(ReactRouterServerErrors.MATCH_NOT_FOUND);
            return done(err);
          }
        });
      }
      else {
        // path utility to make path string compatible in different OS
        // ------------------------------------------------------------
        // use `path.normalize()` to normalzie absolute view file path and absolute base directory path
        // to prevent path strings like `/folder1/folder2/../../folder3/exampleFile`
        // then, derive relative view file path
        // and replace backslash with slash to be compatible on Windows
        data.__meta.view = path.normalize(thing)
          .replace(path.normalize(options.settings.views), '').substring(1)
          .replace('\\', '/');

        var view = require(thing);

        // Check for an ES6 `default` property on the module export
        // ------------------------------------------------------------
        // TypeScript and Babel users that leverage ES6 module depend on this
        // e.g. `export default function MyView() {};`
        if (view.default) {
          view = view.default;
        }

        // create the Component using react's createFactory
        var component = React.createFactory(view);
        return done(null, renderAndDecorate(component(data), data, html));
      }
    }
    catch (err) {

      // on error, pass to the next
      // middleware in the chain!
      return done(err);
    }
  };
};
